### DISCUSSION TITLE
Use pyright to determine a variable's source module without running the Python code?

### DISCUSSION THREAD

**Discussion by xiaohk** [microsoft/pyright#4301] [Q&A] [answered]
Hello, I'm wondering if I can use `pyright` to determine the source module of a variable in Python without running the code?

I have a variable in my Python code and I want to know which module it is from. However, I don't want to run the code in order to find out. I'm looking for a way to statically analyze the code and determine the source module of the variable.

I've tried using tools like `pylint` and `pyflakes`, but they don't seem to provide this information. Is there a way to accomplish this with `pyright`? I feel it is possible, because it seems VSCode Python IntelliSense can figure out the type and info of different variables through static analysis. Unfortunately, I'm not sure how to do it with `pyright` in the command line.

Here is a more concrete example.

```python
import module1
from module2 import item2
import module3 as my_module3
from module4 import item4 as my_item4
from module5 import *

var1 = module1.func_1
var2 = item2
var3 = my_module3.func_3
var4 = my_item4
var5 = func_5
```

Can I use `pyright` in my command line / with a Python script to determine the source module of `var1`...`var5`? Thank you!

**Comment by erictraut** [comment, marked as answer]
Variables are always local to the module in which they are defined. Your code sample above defines a module. Like all modules, it has its own global namespace. There are 10 statements in your sample module, and these statements populate this namespace with the following names: `module1`, `item2`, `my_module3`, `my_item4`, (some indeterminate number of names as defined by `module5`), `var1`, `var2`, `var3`, `var4`, and `var5`. 

Variables can be assigned values using a number of different statements. For example, here are seven different statements that all assign values to the same variable `x`.

```python
import x
import module1 as x
from module1 import x
from module2 import foo as x
x = 1
for x in range(1): pass
with foo() as x: pass
```

Based on your question, I think what you're probably asking is: If a variable is assigned a value using an import statement, is there a way to follow that import back to its source? And if the exported variable in that source is also assigned a value using an import statement, is there a way to follow that back to its source, and continue following that chain of assignments until a variable is discovered that is assigned its value using some means other than an import statement?

Pyright has internal logic that does the above. It is effectively exposed via the "Go To Declarations" language server command. You can invoke this from your editor or call it programmatically.

**Comment by xiaohk** [comment]
Hello @erictraut, thank you so much for your reply!

>Based on your question, I think what you're probably asking is: If a variable is assigned a value using an import statement, is there a way to follow that import back to its source? And if the exported variable in that source is also assigned a value using an import statement, is there a way to follow that back to its source, and continue following that chain of assignments until a variable is discovered that is assigned its value using some means other than an import statement?

Yes, my question is as your described but backward. I want to trace back variable assignments from a particular statement to identify its dependent module(s).

To give a more concrete example, 

```python
from module1 import Class1

my_obj = Class1(args)
my_obj.method()
```

I am interested in the statement `my_obj.method()`, and I want to have a way to programmatically figure out what module is this statement depends on (module `module1`).

Can you give me more pointers on how to invoke the "Go To Declarations" feature programmatically? Thank you again! 🙏

**Reply by erictraut** [reply]
Pyright ships as a command-line tool, a VS Code plugin, and a language server. You will need to use the language server if you want to invoke it programmatically. Refer to [this documentation](https://microsoft.github.io/language-server-protocol/) for details on the language server protocol. If you want to follow sample code, you could look at one or more of the [LSP plugins](https://github.com/Microsoft/pyright#vim) that have been built on top of pyright.

**Reply by xiaohk** [reply]
Got it. Thank you!!
